File: 20220222-java-virtual-storage-design-kata/src/main/java/com/katas/Directory.java
package com.katas;

import java.util.HashSet;
import java.util.Set;

public class Directory extends Resource {

    private final Set<Resource> resources;

    public Directory(String name) {
        super(name, 0);
        this.resources = new HashSet<>();
    }

    public Set<Resource> resources() {
        return new HashSet<>(resources);
    }

    public void add(Resource resource) {
        resources.add(resource);
    }

    @Override
    public int getSize() {
        return resources.stream()
                .mapToInt(Resource::getSize)
                .sum();
    }
}

File: 20200811-kotlin-virtual-storage/src/main/kotlin/com/katas/Directory.kt
package com.katas

data class Directory(
        override val name: String
): Resource {
    private val resources = mutableListOf<Resource>()

    fun list(): List<Resource> {
        return resources.toList()
    }

    fun add(resource: Resource) {
        resources.add(resource)
    }
}
